// number-flow/dist/index.mjs
var ut = Object.defineProperty;
var Q = (n) => {
  throw TypeError(n);
};
var mt = (n, i, t) => i in n ? ut(n, i, { enumerable: !0, configurable: !0, writable: !0, value: t }) : n[i] = t;
var u = (n, i, t) => mt(n, typeof i != "symbol" ? i + "" : i, t), tt = (n, i, t) => i.has(n) || Q("Cannot " + t);
var o = (n, i, t) => (tt(n, i, "read from private field"), t ? t.call(n) : i.get(n)), r = (n, i, t) => i.has(n) ? Q("Cannot add the same private member more than once") : i instanceof WeakSet ? i.add(n) : i.set(n, t), h = (n, i, t, s) => (tt(n, i, "write to private field"), s ? s.call(n, t) : i.set(n, t), t);
import { BROWSER as Y } from "esm-env";
const y = (n, i, t) => {
  const s = document.createElement(n), [e, a] = Array.isArray(i) ? [void 0, i] : [i, t];
  return e && Object.assign(s, e), a == null || a.forEach((l) => s.appendChild(l)), s;
}, gt = (n, i) => {
  typeof Element.prototype.replaceChildren > "u" ? (n.innerHTML = "", i.forEach((t) => n.appendChild(t))) : n.replaceChildren(...i);
}, yt = (n, i) => {
  var t;
  return i === "left" ? n.offsetLeft : (((t = n.offsetParent instanceof HTMLElement ? n.offsetParent : null) == null ? void 0 : t.offsetWidth) ?? 0) - n.offsetWidth - n.offsetLeft;
};
function wt(n, i, { reverse: t = !1 } = {}) {
  const s = n.length;
  for (let e = t ? s - 1 : 0; t ? e >= 0 : e < s; t ? e-- : e++)
    i(n[e], e);
}
function Et(n, i) {
  const t = i.formatToParts(n), s = [], e = [], a = [], l = [], f = {}, c = (p) => (f[p] || (f[p] = 0), `${p}:${f[p]++}`);
  let d = "", x = !1, U = !1;
  for (const p of t) {
    d += p.value;
    const m = p.type === "minusSign" || p.type === "plusSign" ? "sign" : p.type;
    m === "integer" ? (x = !0, e.push(...p.value.split("").map((q) => ({ type: m, value: parseInt(q) })))) : m === "group" ? e.push({ type: m, value: p.value }) : m === "decimal" ? (U = !0, a.push({ type: m, value: p.value, key: c(m) })) : m === "fraction" ? a.push(...p.value.split("").map((q) => ({ type: m, value: parseInt(q), key: c(m) }))) : (x || U ? l : s).push({
      type: m,
      value: p.value,
      key: c(m)
    });
  }
  const _ = [];
  for (let p = e.length - 1; p >= 0; p--)
    _.unshift({ ...e[p], key: c(e[p].type) });
  return { pre: s, integer: _, fraction: a, post: l, formatted: d };
}
const xt = Y ? HTMLElement : class {
}, _t = String.raw, bt = Y && typeof CSS < "u" && CSS.supports("transition-timing-function", "linear(1, 2)"), rt = Y && typeof CSS < "u" && typeof CSS.registerProperty < "u", et = Y && typeof CSS < "u" && typeof CSS.supports("animation-composition", "accumulate") < "u";
if (rt)
  try {
    CSS.registerProperty({
      name: "--_number-flow-scale-x",
      syntax: "<number>",
      inherits: !1,
      initialValue: "1"
    });
  } catch {
  }
const ct = "var(--number-flow-char-height, 1em)", g = "var(--number-flow-mask-height, 0.25em)", E = "var(--number-flow-mask-width, 0.5em)", Z = `${E} * 1/var(--_number-flow-scale-x)`, it = `calc(${Z})`, st = `linear-gradient(to bottom, transparent 0, #000 ${g}, #000 calc(100% - ${g}), transparent 100%)`, K = "#000 0, transparent 71%", Ct = "span", Tt = {
  fontKerning: "none",
  display: "inline-block",
  lineHeight: ct,
  padding: `${g} 0`
}, nt = _t`:host{display:inline-block;direction:ltr;user-select:none;pointer-events:none;white-space:nowrap;position:relative;line-height:${ct} !important;isolation:isolate;}::slotted(${Ct}){position:absolute;left:0;top:0;color:transparent !important;z-index:-1;user-select:text;pointer-events:auto;}.section{display:inline-block;padding-bottom:${g};padding-top:${g};vertical-align:top;user-select:none;}.section:not(.section--masked),.section--masked .section__inner{position:relative;isolation:isolate;}.section__inner{display:inline-block;transform-origin:inherit;}.section--justify-left{transform-origin:center left;}.section--justify-right{transform-origin:center right;}.section--masked{overflow:clip;position:relative;z-index:-1;--_number-flow-scale-x:1;-webkit-mask-repeat:no-repeat;-webkit-mask-size:calc(100% - ${Z}) 100%,100% calc(100% - ${g} * 2),${it} ${g},${it} ${g};}.section--masked.section--justify-left{padding-right:${E};margin-right:calc(-1 * ${E});-webkit-mask-image:${st},linear-gradient(to right,#000 0,#000 calc(100% - ${Z}),transparent 100%),radial-gradient(at bottom left,${K}),radial-gradient(at top left,${K});-webkit-mask-position:left,center,right top,right bottom;}.section--masked.section--justify-right{padding-left:${E};margin-left:calc(-1 * ${E});-webkit-mask-image:${st},linear-gradient(to left,#000 0,#000 calc(100% - ${Z}),transparent 100%),radial-gradient(at bottom right,${K}),radial-gradient(at top right,${K});-webkit-mask-position:right,center,left top,left bottom;}.section__exiting{position:absolute !important;z-index:-1;}.digit{display:inline-block;position:relative;}.digit__stack{display:flex;flex-direction:column;align-items:center;gap:${g};position:absolute;width:100%;}.digit__stack > *{display:inline-block;}.digit__lt{bottom:calc(100% + ${g});}.digit__gt{top:calc(100% + ${g});}.symbol{display:inline-block;position:relative;isolation:isolate;}.symbol__value{display:inline-block;white-space:pre;}.symbol__exiting{position:absolute;z-index:-1;}.section--justify-left .symbol__exiting{left:0;}.section--justify-right .symbol__exiting{right:0;}`;
function kt(n, i, t = 60) {
  const s = Math.trunc(n / 1e3 * t);
  return Array.from({ length: s }, (e, a) => i(a / (s - 1)));
}
function vt(n, i, t = 60) {
  if (rt)
    return [i(0), i(1)];
  const s = Math.max(Math.trunc(n / 1e3 * t), 0), e = Array.from({ length: s }, (a, l) => ({
    ...i(l / s),
    offset: (l + 0.5) / s
  }));
  return e.unshift({
    ...i(0),
    offset: 0
  }), e;
}
const ot = (n, i, t) => n + (i - n) * t, $t = { duration: 500, easing: "ease-out" }, ht = bt ? {
  duration: 900,
  easing: "linear(0,.005,.019,.039,.066,.096,.129,.165,.202,.24,.278,.316,.354,.39,.426,.461,.494,.526,.557,.586,.614,.64,.665,.689,.711,.731,.751,.769,.786,.802,.817,.831,.844,.856,.867,.877,.887,.896,.904,.912,.919,.925,.931,.937,.942,.947,.951,.955,.959,.962,.965,.968,.971,.973,.976,.978,.98,.981,.983,.984,.986,.987,.988,.989,.99,.991,.992,.992,.993,.994,.994,.995,.995,.996,.996,.9963,.9967,.9969,.9972,.9975,.9977,.9979,.9981,.9982,.9984,.9985,.9987,.9988,.9989,1)"
} : {
  duration: 900,
  // Spring-like cubic-bezier stolen from Vaul: https://vaul.emilkowal.ski/
  easing: "cubic-bezier(.32,.72,0,1)"
}, St = ht;
let G;
var T, b, C, k, v;
class Mt extends xt {
  constructor() {
    super(...arguments);
    u(this, "xTiming", ht);
    u(this, "yTiming", St);
    u(this, "fadeTiming", $t);
    u(this, "root", !1);
    u(this, "manual", !1);
    r(this, T, !1);
    r(this, b);
    r(this, C);
    r(this, k);
    r(this, v);
  }
  static define() {
    Y && customElements.define("number-flow", this);
  }
  set parts(t) {
    if (t == null)
      return;
    const { pre: s, integer: e, fraction: a, post: l } = t;
    if (o(this, T))
      this.manual || this.willUpdate(), o(this, b).update(s), o(this, C).update(e), o(this, k).update(a), o(this, v).update(l), this.manual || this.didUpdate();
    else {
      if (this.attachShadow({ mode: "open" }), typeof CSSStyleSheet < "u" && this.shadowRoot.adoptedStyleSheets)
        G || (G = new CSSStyleSheet(), G.replaceSync(nt)), this.shadowRoot.adoptedStyleSheets = [G];
      else {
        const f = document.createElement("style");
        f.textContent = nt, this.shadowRoot.appendChild(f);
      }
      this.shadowRoot.appendChild(y("slot")), h(this, b, new lt(this, s, {
        // part: 'pre',
        inert: !0,
        ariaHidden: "true",
        justify: "right"
      })), this.shadowRoot.appendChild(o(this, b).el), h(this, C, new at(this, e, {
        // part: 'integer',
        inert: !0,
        ariaHidden: "true",
        justify: "right"
      })), this.shadowRoot.appendChild(o(this, C).el), h(this, k, new at(this, a, {
        // part: 'fraction',
        inert: !0,
        ariaHidden: "true",
        justify: "left"
      })), this.shadowRoot.appendChild(o(this, k).el), h(this, v, new lt(this, l, {
        // part: 'post',
        inert: !0,
        ariaHidden: "true",
        justify: "left"
      })), this.shadowRoot.appendChild(o(this, v).el);
    }
    h(this, T, !0);
  }
  willUpdate() {
    const t = this.root ? this.getBoundingClientRect() : new DOMRect();
    o(this, b).willUpdate(t), o(this, C).willUpdate(t), o(this, k).willUpdate(t), o(this, v).willUpdate(t);
  }
  didUpdate() {
    const t = this.root ? this.getBoundingClientRect() : new DOMRect();
    o(this, b).didUpdate(t), o(this, C).didUpdate(t), o(this, k).didUpdate(t), o(this, v).didUpdate(t);
  }
}
T = new WeakMap(), b = new WeakMap(), C = new WeakMap(), k = new WeakMap(), v = new WeakMap();
class dt {
  constructor(i, t, { justify: s, className: e, ...a }, l) {
    u(this, "flow");
    u(this, "el");
    u(this, "justify");
    // All children in the DOM:
    u(this, "children", /* @__PURE__ */ new Map());
    u(this, "onCharRemove", (i) => () => {
      this.children.delete(i);
    });
    this.flow = i, this.justify = s;
    const f = t.map((c) => this.addChar(c).el);
    f.push(document.createTextNode("​")), this.el = y("span", {
      ...a,
      className: `section section--justify-${s} ${e ?? ""}`
    }, l ? l(f) : f);
  }
  addChar(i, { startDigitsAtZero: t = !1, ...s } = {}) {
    const e = i.type === "integer" || i.type === "fraction" ? new pt(this, i.type, t ? 0 : i.value, {
      ...s,
      onRemove: this.onCharRemove(i.key)
    }) : new jt(this, i.type, i.value, {
      ...s,
      onRemove: this.onCharRemove(i.key)
    });
    return this.children.set(i.key, e), e;
  }
  unpop(i) {
    i.el.classList.remove("section__exiting"), i.el.style[this.justify] = "";
  }
  pop(i) {
    i.forEach((t) => {
      t.el.style[this.justify] = `${yt(t.el, this.justify)}px`;
    }), i.forEach((t) => {
      t.el.classList.add("section__exiting"), t.present = !1;
    });
  }
  addNewAndUpdateExisting(i, t = this.el) {
    const s = /* @__PURE__ */ new Map(), e = /* @__PURE__ */ new Map(), a = this.justify === "left", l = a ? "prepend" : "append";
    wt(i, (c) => {
      let d;
      this.children.has(c.key) ? (d = this.children.get(c.key), e.set(c, d), this.unpop(d), d.present = !0) : (d = this.addChar(c, { startDigitsAtZero: !0, animateIn: !0 }), s.set(c, d)), t[l](d.el);
    }, { reverse: a });
    const f = t.getBoundingClientRect();
    s.forEach((c) => {
      c.willUpdate(f);
    }), s.forEach((c, d) => {
      c.update(d.value);
    }), e.forEach((c, d) => {
      c.update(d.value);
    });
  }
}
var $, M, A, N, B, L;
class at extends dt {
  constructor(t, s, { className: e, ...a }) {
    let l;
    super(t, s, {
      ...a,
      className: `${e ?? ""} section--masked`
    }, (f) => [
      l = y("span", {
        className: "section__inner"
      }, f)
    ]);
    r(this, $);
    r(this, M);
    r(this, A);
    r(this, N);
    r(this, B);
    r(this, L);
    h(this, $, l);
  }
  willUpdate(t) {
    const s = this.el.getBoundingClientRect();
    h(this, M, s.width), h(this, A, s[this.justify] - t[this.justify]);
    const e = o(this, $).getBoundingClientRect();
    this.children.forEach((a) => a.willUpdate(e));
  }
  update(t) {
    const s = /* @__PURE__ */ new Map();
    this.children.forEach((e, a) => {
      t.find((l) => l.key === a) || s.set(a, e), this.unpop(e);
    }), this.addNewAndUpdateExisting(t, o(this, $)), s.forEach((e) => {
      e instanceof pt && e.update(0);
    }), this.pop(s);
  }
  didUpdate(t) {
    var c, d, x, U;
    (c = o(this, N)) == null || c.cancel(), (d = o(this, L)) == null || d.cancel(), (x = o(this, B)) == null || x.cancel();
    const s = this.el.getBoundingClientRect(), e = s[this.justify] - t[this.justify], a = o(this, A) - e, l = Math.max(o(this, M), 0.01) / Math.max(s.width, 0.01), f = o(this, $).getBoundingClientRect();
    this.children.forEach((_) => _.didUpdate(f)), h(this, N, this.el.animate({
      transform: [`translateX(${a}px) scaleX(${l})`, "none"]
    }, this.flow.xTiming)), l !== 1 && (h(this, L, (U = o(this, $)) == null ? void 0 : U.animate({
      // 1/x isn't linear so we need to do sampling:
      transform: kt(1e3, (_) => `scaleX(${1 / ot(l, 1, _)})`)
    }, this.flow.xTiming)), h(this, B, this.el.animate(vt(1e3, (_) => ({
      "--_number-flow-scale-x": ot(l, 1, _)
    })), this.flow.xTiming)));
  }
}
$ = new WeakMap(), M = new WeakMap(), A = new WeakMap(), N = new WeakMap(), B = new WeakMap(), L = new WeakMap();
var H, P;
class lt extends dt {
  constructor() {
    super(...arguments);
    r(this, H);
    r(this, P);
  }
  willUpdate(t) {
    const s = this.el.getBoundingClientRect();
    h(this, H, s[this.justify] - t[this.justify]), this.children.forEach((e) => e.willUpdate(s));
  }
  update(t) {
    const s = /* @__PURE__ */ new Map();
    this.children.forEach((e, a) => {
      t.find((l) => l.key === a) || s.set(a, e);
    }), this.pop(s), this.addNewAndUpdateExisting(t);
  }
  didUpdate(t) {
    var a;
    (a = o(this, P)) == null || a.cancel();
    const s = this.el.getBoundingClientRect(), e = s[this.justify] - t[this.justify];
    this.children.forEach((l) => l.didUpdate(s)), h(this, P, this.el.animate({
      transform: [`translateX(${o(this, H) - e}px)`, "none"]
    }, this.flow.xTiming));
  }
}
H = new WeakMap(), P = new WeakMap();
var j, W, R;
class J {
  constructor(i, t, { onRemove: s, animateIn: e = !1 } = {}) {
    u(this, "flow");
    u(this, "el");
    r(this, j, !0);
    r(this, W);
    r(this, R);
    this.flow = i, this.el = t, e && this.el.animate({
      opacity: ["0", "1"]
    }, i.fadeTiming), h(this, W, s);
  }
  get present() {
    return o(this, j);
  }
  set present(i) {
    var s;
    if (o(this, j) === i)
      return;
    const t = getComputedStyle(this.el).getPropertyValue("opacity");
    (s = o(this, R)) == null || s.cancel(), h(this, R, this.el.animate({
      opacity: [t, i ? "1" : "0"]
    }, this.flow.fadeTiming)), i || (o(this, R).onfinish = () => {
      var e;
      this.el.remove(), (e = o(this, W)) == null || e.call(this);
    }), h(this, j, i);
  }
}
j = new WeakMap(), W = new WeakMap(), R = new WeakMap();
class ft extends J {
  constructor(t, s, e, a) {
    super(t.flow, e, a);
    u(this, "section");
    u(this, "value");
    u(this, "el");
    this.section = t, this.value = s, this.el = e;
  }
}
var S, X, z, O, D;
class pt extends ft {
  constructor(t, s, e, a) {
    super(t, e, y("span", {
      className: "digit",
      // part: `digit ${type} ${value}`,
      textContent: e + ""
    }), a);
    r(this, S);
    // Relative to parent:
    r(this, X);
    r(this, z);
    r(this, O);
    r(this, D);
  }
  willUpdate(t) {
    h(this, S, this.value);
    const s = this.el.getBoundingClientRect();
    h(this, X, s.y - t.y);
    const e = s[this.section.justify] - t[this.section.justify], a = s.width / 2;
    h(this, z, this.section.justify === "left" ? e + a : e - a);
  }
  update(t) {
    this.value = t, o(this, S) !== t && gt(this.el, [
      ...t === 0 ? [] : [
        y("span", { className: "digit__stack digit__lt" }, Array.from({ length: t }, (s, e) => y("span", { textContent: e + "" })))
      ],
      document.createTextNode(t + ""),
      ...t === 9 ? [] : [
        y("span", { className: "digit__stack digit__gt" }, Array.from({ length: 9 - t }, (s, e) => y("span", { textContent: t + e + 1 + "" })))
      ]
    ]);
  }
  didUpdate(t) {
    var d, x;
    (d = o(this, O)) == null || d.cancel(), o(this, S) !== this.value && ((x = o(this, D)) == null || x.cancel());
    const s = this.el.getBoundingClientRect(), e = s[this.section.justify] - t[this.section.justify], a = s.width / 2, l = this.section.justify === "left" ? e + a : e - a, f = `translateX(${o(this, z) - l}px)`, c = `translateY(calc((100% + ${g}) * ${this.value - o(this, S)} + ${o(this, X)}px))`;
    h(this, O, this.el.animate({
      transform: [et ? f : `${f} ${c}`, "none"]
    }, {
      ...this.flow.xTiming,
      composite: "accumulate"
      // in case there's an in-flight y animation
    })), et && o(this, S) !== this.value && h(this, D, this.el.animate({
      transform: [c, "none"]
    }, {
      ...this.flow.yTiming,
      composite: "accumulate"
    }));
  }
}
S = new WeakMap(), X = new WeakMap(), z = new WeakMap(), O = new WeakMap(), D = new WeakMap();
var w, I, F, V;
class jt extends ft {
  constructor(t, s, e, a) {
    const l = y("span", {
      className: "symbol__value",
      textContent: e
    });
    super(t, e, y("span", {
      className: "symbol"
    }, [l]), a);
    u(this, "type");
    r(this, w, /* @__PURE__ */ new Map());
    r(this, I);
    r(this, F, (t) => () => {
      o(this, w).delete(t);
    });
    r(this, V);
    this.type = s, o(this, w).set(e, new J(this.flow, l, {
      onRemove: o(this, F).call(this, e)
    }));
  }
  willUpdate(t) {
    if (this.type === "decimal")
      return;
    const s = this.el.getBoundingClientRect();
    h(this, I, s[this.section.justify] - t[this.section.justify]);
  }
  update(t) {
    if (this.value !== t) {
      const s = o(this, w).get(this.value);
      if (s.present = !1, s.el.classList.add("symbol__exiting"), o(this, w).has(t)) {
        const e = o(this, w).get(t);
        e.present = !0, e.el.classList.remove("symbol__exiting");
      } else {
        const e = y("span", {
          className: "symbol__value",
          textContent: t
        });
        this.el.appendChild(e), o(this, w).set(t, new J(this.flow, e, {
          animateIn: !0,
          onRemove: o(this, F).call(this, t)
        }));
      }
    }
    this.value = t;
  }
  didUpdate(t) {
    var a;
    if (this.type === "decimal")
      return;
    (a = o(this, V)) == null || a.cancel();
    const e = this.el.getBoundingClientRect()[this.section.justify] - t[this.section.justify];
    h(this, V, this.el.animate({
      transform: [`translateX(${o(this, I) - e}px)`, "none"]
    }, this.flow.xTiming));
  }
}
w = new WeakMap(), I = new WeakMap(), F = new WeakMap(), V = new WeakMap();
export {
  Mt as NumberFlowLite,
  Ct as SlottedTag,
  $t as defaultFadeTiming,
  ht as defaultXTiming,
  St as defaultYTiming,
  Et as partitionParts,
  Tt as slottedStyles,
  et as supportsAnimationComposition,
  bt as supportsLinear
};
