// @number-flow/react/dist/framer-motion.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/framer-motion.ts
var framer_motion_exports = {};
__export(framer_motion_exports, {
  toEase: () => toEase,
  toTransition: () => toTransition,
  useTiming: () => useTiming
});
module.exports = __toCommonJS(framer_motion_exports);
var React = __toESM(require("react"));
var import_framer_motion = require("framer-motion");
var bezierRegExp = /^cubic-bezier\((.*?),(.*?),(.*?),(.*?)\)$/;
var linearRegExp = /^linear\((.*?)\)$/;
function toEase(_easing) {
  const easing = _easing.trim();
  if (easing === "linear") return easing;
  if (easing === "ease-in") return "easeIn";
  if (easing === "ease-out") return "easeOut";
  if (easing === "ease-in-out") return "easeInOut";
  const bezierMatch = easing.match(bezierRegExp);
  if (bezierMatch) {
    return bezierMatch.slice(1).map(parseFloat);
  }
  const linearMatch = easing.match(linearRegExp);
  const linearExpr = linearMatch?.[1];
  if (linearExpr) {
    const points = linearExpr.split(",");
    const times = [];
    const values = [];
    try {
      points.forEach((point, i) => {
        const [p1, p2] = point.trim().split(/\s+/);
        if (p1?.includes("%")) {
          times.push(parseFloat(p1) / 100);
          values.push(parseFloat(p2));
        } else if (p2?.includes("%")) {
          times.push(parseFloat(p2) / 100);
          values.push(parseFloat(p1));
        } else {
          times.push(i / (points.length - 1));
          values.push(parseFloat(p1));
        }
      });
    } catch {
      throw new Error(`Cannot parse linear() expression: ${easing}`);
    }
    return (0, import_framer_motion.transform)(times, values);
  }
  throw new Error(`Cannot convert ${easing} to a Framer Motion easing`);
}
function toTransition(timing) {
  const { duration = 0, delay = 0, easing = "ease" } = timing;
  if (typeof duration != "number") throw new Error("Duration must be a number");
  return {
    duration: duration / 1e3,
    delay: delay / 1e3,
    ease: toEase(easing)
  };
}
function useTiming(timing) {
  const [ret, setRet] = React.useState(toTransition(timing));
  React.useEffect(() => {
    setRet(toTransition(timing));
  }, [timing]);
  return ret;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  toEase,
  toTransition,
  useTiming
});
