// @number-flow/react/dist/framer-motion.mjs
// src/framer-motion.ts
import * as React from "react";
import { transform } from "framer-motion";
var bezierRegExp = /^cubic-bezier\((.*?),(.*?),(.*?),(.*?)\)$/;
var linearRegExp = /^linear\((.*?)\)$/;
function toEase(_easing) {
  const easing = _easing.trim();
  if (easing === "linear") return easing;
  if (easing === "ease-in") return "easeIn";
  if (easing === "ease-out") return "easeOut";
  if (easing === "ease-in-out") return "easeInOut";
  const bezierMatch = easing.match(bezierRegExp);
  if (bezierMatch) {
    return bezierMatch.slice(1).map(parseFloat);
  }
  const linearMatch = easing.match(linearRegExp);
  const linearExpr = linearMatch?.[1];
  if (linearExpr) {
    const points = linearExpr.split(",");
    const times = [];
    const values = [];
    try {
      points.forEach((point, i) => {
        const [p1, p2] = point.trim().split(/\s+/);
        if (p1?.includes("%")) {
          times.push(parseFloat(p1) / 100);
          values.push(parseFloat(p2));
        } else if (p2?.includes("%")) {
          times.push(parseFloat(p2) / 100);
          values.push(parseFloat(p1));
        } else {
          times.push(i / (points.length - 1));
          values.push(parseFloat(p1));
        }
      });
    } catch {
      throw new Error(`Cannot parse linear() expression: ${easing}`);
    }
    return transform(times, values);
  }
  throw new Error(`Cannot convert ${easing} to a Framer Motion easing`);
}
function toTransition(timing) {
  const { duration = 0, delay = 0, easing = "ease" } = timing;
  if (typeof duration != "number") throw new Error("Duration must be a number");
  return {
    duration: duration / 1e3,
    delay: delay / 1e3,
    ease: toEase(easing)
  };
}
function useTiming(timing) {
  const [ret, setRet] = React.useState(toTransition(timing));
  React.useEffect(() => {
    setRet(toTransition(timing));
  }, [timing]);
  return ret;
}
export {
  toEase,
  toTransition,
  useTiming
};
