// @number-flow/react/dist/index.mjs
"use client";

// src/index.tsx
import * as React from "react";
import {
  SlottedTag,
  slottedStyles,
  partitionParts,
  NumberFlowLite,
  supportsLinear
} from "number-flow";
import { defaultXTiming, defaultYTiming } from "number-flow";
var OBSERVED_ATTRIBUTES = ["parts"];
var NumberFlowElement = class extends NumberFlowLite {
  static observedAttributes = OBSERVED_ATTRIBUTES;
  attributeChangedCallback(attr, _oldValue, newValue) {
    this[attr] = JSON.parse(newValue);
  }
};
NumberFlowElement.define();
var formatters = {};
var NumberFlowPriv = class extends React.Component {
  constructor(props) {
    super(props);
    this.handleRef = this.handleRef.bind(this);
  }
  // Update the non-parts props to avoid JSON serialization
  // Parts needs to be set in render still:
  updateNonPartsProps() {
    if (this.#el) {
      if (this.props.root != null) this.#el.root = this.props.root;
      if (this.props.fadeTiming) this.#el.fadeTiming = this.props.fadeTiming;
      if (this.props.xTiming) this.#el.xTiming = this.props.xTiming;
      if (this.props.yTiming) this.#el.yTiming = this.props.yTiming;
    }
  }
  componentDidMount() {
    this.updateNonPartsProps();
    if (this.#el) this.#el.manual = true;
  }
  getSnapshotBeforeUpdate() {
    this.updateNonPartsProps();
    this.#el?.willUpdate();
    return null;
  }
  componentDidUpdate() {
    this.#el?.didUpdate();
  }
  #el;
  handleRef(el) {
    if (this.props.innerRef) this.props.innerRef.current = el;
    this.#el = el;
  }
  render() {
    const {
      innerRef,
      className,
      parts,
      // These are set in updateNonPartsProps, so ignore them here:
      fadeTiming,
      xTiming,
      yTiming,
      ...rest
    } = this.props;
    return (
      // @ts-expect-error missing types
      /* @__PURE__ */ React.createElement(
        "number-flow",
        {
          ref: this.handleRef,
          class: className,
          ...rest,
          parts: JSON.stringify(parts)
        },
        /* @__PURE__ */ React.createElement(SlottedTag, { style: slottedStyles }, parts.formatted)
      )
    );
  }
};
var NumberFlow = React.forwardRef(function NumberFlow2({ value, locales, format, ...props }, _ref) {
  React.useImperativeHandle(_ref, () => ref.current, []);
  const ref = React.useRef();
  const localesString = React.useMemo(() => locales ? JSON.stringify(locales) : "", [locales]);
  const formatString = React.useMemo(() => format ? JSON.stringify(format) : "", [format]);
  const parts = React.useMemo(() => {
    const formatter = formatters[`${localesString}:${formatString}`] ??= new Intl.NumberFormat(
      locales,
      format
    );
    return partitionParts(value, formatter);
  }, [value, localesString, formatString]);
  return /* @__PURE__ */ React.createElement(NumberFlowPriv, { ...props, parts, innerRef: ref });
});
var src_default = NumberFlow;
function useSupportsLinear() {
  const [supported, setSupported] = React.useState(false);
  React.useEffect(() => {
    setSupported(supportsLinear);
  }, []);
  return supported;
}
function useLinear(linear, fallback) {
  const supported = useSupportsLinear();
  return supported ? linear : fallback;
}
export {
  NumberFlowElement,
  src_default as default,
  defaultXTiming,
  defaultYTiming,
  useLinear,
  useSupportsLinear
};
