// @number-flow/react/dist/index.js
"use strict";
"use client";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  NumberFlowElement: () => NumberFlowElement,
  default: () => src_default,
  defaultXTiming: () => import_number_flow2.defaultXTiming,
  defaultYTiming: () => import_number_flow2.defaultYTiming,
  useLinear: () => useLinear,
  useSupportsLinear: () => useSupportsLinear
});
module.exports = __toCommonJS(src_exports);
var React = __toESM(require("react"));
var import_number_flow = require("number-flow");
var import_number_flow2 = require("number-flow");
var OBSERVED_ATTRIBUTES = ["parts"];
var NumberFlowElement = class extends import_number_flow.NumberFlowLite {
  static observedAttributes = OBSERVED_ATTRIBUTES;
  attributeChangedCallback(attr, _oldValue, newValue) {
    this[attr] = JSON.parse(newValue);
  }
};
NumberFlowElement.define();
var formatters = {};
var NumberFlowPriv = class extends React.Component {
  constructor(props) {
    super(props);
    this.handleRef = this.handleRef.bind(this);
  }
  // Update the non-parts props to avoid JSON serialization
  // Parts needs to be set in render still:
  updateNonPartsProps() {
    if (this.#el) {
      if (this.props.root != null) this.#el.root = this.props.root;
      if (this.props.fadeTiming) this.#el.fadeTiming = this.props.fadeTiming;
      if (this.props.xTiming) this.#el.xTiming = this.props.xTiming;
      if (this.props.yTiming) this.#el.yTiming = this.props.yTiming;
    }
  }
  componentDidMount() {
    this.updateNonPartsProps();
    if (this.#el) this.#el.manual = true;
  }
  getSnapshotBeforeUpdate() {
    this.updateNonPartsProps();
    this.#el?.willUpdate();
    return null;
  }
  componentDidUpdate() {
    this.#el?.didUpdate();
  }
  #el;
  handleRef(el) {
    if (this.props.innerRef) this.props.innerRef.current = el;
    this.#el = el;
  }
  render() {
    const {
      innerRef,
      className,
      parts,
      // These are set in updateNonPartsProps, so ignore them here:
      fadeTiming,
      xTiming,
      yTiming,
      ...rest
    } = this.props;
    return (
      // @ts-expect-error missing types
      /* @__PURE__ */ React.createElement(
        "number-flow",
        {
          ref: this.handleRef,
          class: className,
          ...rest,
          parts: JSON.stringify(parts)
        },
        /* @__PURE__ */ React.createElement(import_number_flow.SlottedTag, { style: import_number_flow.slottedStyles }, parts.formatted)
      )
    );
  }
};
var NumberFlow = React.forwardRef(function NumberFlow2({ value, locales, format, ...props }, _ref) {
  React.useImperativeHandle(_ref, () => ref.current, []);
  const ref = React.useRef();
  const localesString = React.useMemo(() => locales ? JSON.stringify(locales) : "", [locales]);
  const formatString = React.useMemo(() => format ? JSON.stringify(format) : "", [format]);
  const parts = React.useMemo(() => {
    const formatter = formatters[`${localesString}:${formatString}`] ??= new Intl.NumberFormat(
      locales,
      format
    );
    return (0, import_number_flow.partitionParts)(value, formatter);
  }, [value, localesString, formatString]);
  return /* @__PURE__ */ React.createElement(NumberFlowPriv, { ...props, parts, innerRef: ref });
});
var src_default = NumberFlow;
function useSupportsLinear() {
  const [supported, setSupported] = React.useState(false);
  React.useEffect(() => {
    setSupported(import_number_flow.supportsLinear);
  }, []);
  return supported;
}
function useLinear(linear, fallback) {
  const supported = useSupportsLinear();
  return supported ? linear : fallback;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  NumberFlowElement,
  defaultXTiming,
  defaultYTiming,
  useLinear,
  useSupportsLinear
});
